cilium/CFP-39876-clustermesh-filtered-export.md

```diff
@@ -0,0 +1,211 @@
+
+Cilium `NetworkPolicy` typically uses selectors like:
+
+```yaml
+endpointSelector:
+  matchLabels:
+    app: frontend
+ingress:
+  - fromEndpoints:
+    - matchLabels:
+        app: backend
+        tier: api
+```
+
+Without labels, the policy engine cannot:
+-Match fromEndpoints selectors
+-Evaluate matchLabels conditions
+-Distinguish between different application identities
+
+
+When using this feature, cross-cluster communication requires both the source and destination namespaces to be global, and both clusters to have an established cluster mesh. The evaluation of network policy for different cases is summarized below:
+
+### Summary of Enforcement Logic
+
+| Source Pod in Global NS | Destination Pod in Global NS | Can you allow traffic based on Labels? |
+|-------------------------|------------------------------|------------------|
+| Yes                     | Yes                          | ✅ Enforced normally |
+| No                      | Yes                          | ❌ Not Supported  |
+| Yes                     | No                           | ❌ Not Supported |
+| No                      | No                           | ❌ Not Supported  |
+
+In the above table, Not Supported implies the policies are not enforced due to missing source identity or labels information at the destination cluster. 
+
+## Global Service Support
+For a service to be considered a Global Service, it must satisfy the following conditions 
+- It must be annotated with existing global service annotation ***service.cilium.io/global: "true"***
+- It must reside within a global namespace 
+
+## Implementation details
+The proposed namespace based export mode has a config that users can pass in to mark all non annotated namespaces as global
+- clustermesh-default-global-namespace (true/false) - Marks implicitly all non annotated namespaces as global if set to true. Marks implicitly all non annotated namespaces as local if set to false. This option is set to true by default to preserve the current behavior.
+
+#### Allow/Deny listing namespace
+To allowlist a namespace and mark it as global, users can annotate the namespace with:
+
+```
+clustermesh.cilium.io/global: "true"
+```
+
+The ClusterMesh API server will watch namespace resources. Adding, changing or removing this namespace annotation will result in the following steps:
+
+- A full synchronization is triggered for all resources within the newly annotated namespaces 
+- Only entries associated with namespaces explicitly annotated as global will be exported.
+- Unannotated namespaces are treated as global/local depending on the value of config `clustermesh-default-global-namespace`
+
+To explicitly mark a namespace as local, the annotation can be set as:
+
+```
+clustermesh.cilium.io/global: "false"
+```
+
+When the annotation is removed or disabled for a namespace:
+- A full synchronization is triggered to mark all resources within the annotated namespace as global/local
```

Comment:
What is involved in this "mark all resources" step?

When it comes to user-managed resources I think it makes a lot of sense to ensure the user is responsible for setting the annotations. If there are resources managed only by Cilium, I think it makes sense for Cilium to propagate such annotations. What I'm less clear about here is for instance does this mean that if the user adds a namespace annotation, Cilium will perform write operations against Pods and Services in the namespace to propagate the annotations? Is that what this statement means? (If yes, is that really necessary?)

Reply:
Ah right, this is about the forward propagation into etcd :+1: Makes sense yep.